### PR TITLE
Added in a VcfFuncotationFactory.

### LINKED ISSUE #4357: Funcotator - Add dbSnp data source
Must add in a VCF input for data sources.

Once this is done, must add in the dbSnp source.

### PR DESCRIPTION
**Pull request by jonn-smith**
Fixes #4357

Need to release new copy of data sources along with this.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 1]
```diff
@@ -812,7 +813,7 @@ private String createProgramGroupID(final SAMFileHeader header) {
                                                                                   final String name,
                                                                                   final Class<? extends Feature> featureType) {
 
-        final FeatureInput<? extends Feature> featureInput = new FeatureInput<>(name + ":" + filePath);
+        final FeatureInput<? extends Feature> featureInput = new FeatureInput<>(name + FEATURE_ARGUMENT_TAG_DELIMITER + filePath);
```
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Minor... Can you change this to `FeatureInput.FEATURE_ARGUMENT_TAG_DELIMITER` and change the import statement above?

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/DataSourceFuncotationFactory.java [hunk 2]
```diff
@@ -99,14 +100,118 @@ public String getVersion() {
...
+     * For some Data Sources knowledge of Gene Name or Transcript ID is required for annotation.
+     * Accounts for override values passed into the constructor as well.
+     * @param variant {@link VariantContext} to annotate.
+     * @param referenceContext {@link ReferenceContext} corresponding to the given {@code variant}.
+     * @param featureSourceMap {@link Map} of {@link String} -> {@link List} of {@link Feature} (data source name -> data source features corresponding to the given {@code variant}.
+     * @param gencodeFuncotations {@link List} of {@link GencodeFuncotation} that have already been created for the given {@code variant}/{@code referenceContext}/{@code featureContext}.
+     * @return {@link List} of {@link Funcotation} given the {@code variant}, {@code referenceContext}, and {@code featureContext}.  This should never be empty.
+     */
+    public List<Funcotation> createFuncotations(final VariantContext variant, final ReferenceContext referenceContext, final Map<String, List<Feature>> featureSourceMap, final List<GencodeFuncotation> gencodeFuncotations) {
```
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/DataSourceFuncotationFactory.java]: Can you combine this method with the one above to leverage shared code?  If a data source needs gencode info, but `createFuncotation(...)` is called without Gencode, will an exception be thrown?

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorArgumentDefinitions.java [hunk 3]
```diff
@@ -101,7 +101,17 @@ public void assertConfigFilePropertiesAreValid(final Properties configFileProper
         },
 
         /**
-         * This values indicates a GENCODE GTF data file.
+         * This value indicates a Variant Context File (VCF)
+         */
+        VCF("vcf") {
+            @Override
+            public void assertConfigFilePropertiesAreValid(final Properties configFileProperties, final Path configFilePath) {
+                // There is no special check required for vcf files.
```
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorArgumentDefinitions.java]: Can you add a sentence of "why not" ?

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/cosmic/CosmicFuncotationFactory.java [hunk 4]
```diff
@@ -191,13 +197,13 @@ public String getName() {
      * This method should never be called on a {@link CosmicFuncotationFactory} - knowledge of the applied
      * {@link GencodeFuncotation}s is required to create a {@link Funcotation} from here.
      */
-    public List<Funcotation> createFuncotations(final VariantContext variant, final ReferenceContext referenceContext, final List<Feature> featureList) {
+    protected List<Funcotation> createFuncotationsOnVariant(final VariantContext variant, final ReferenceContext referenceContext, final List<Feature> featureList) {
         // TODO: this should be allowed, but with a warning to the user that only annotations with good Genome Positions can be used.
```
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/cosmic/CosmicFuncotationFactory.java]: Did you address this?

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 1 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Sounds good.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/DataSourceFuncotationFactory.java (hunk 2 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/DataSourceFuncotationFactory.java]: Ha!  Yeah - I forgot to combine them.

Yeah - I'll add that in.

Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/DataSourceFuncotationFactory.java]: Yeah - the exception gets thrown in the concrete classes implementations of `createFuncotations`

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorArgumentDefinitions.java (hunk 3 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorArgumentDefinitions.java]: Sure.

File: src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/cosmic/CosmicFuncotationFactory.java (hunk 4 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/cosmic/CosmicFuncotationFactory.java]: Not yet.

Added as issue #4594

**Comment by jonn-smith** [comment]
OK @LeeTL1220 - first round comments addressed.

**Pull request merged**